File: 649*dota2-senate/solution.py
class Solution(object):
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """
        r_lst, d_lst = [], []
        n = len(senate)        
        
        for idx, c in enumerate(senate):
            if c == 'R':
                r_lst.append([idx, c])
            else:
                d_lst.append([idx, c])
                
        while r_lst and d_lst:
            r, d = r_lst.pop(0), d_lst.pop(0)
            r_idx, r_val = r[0], r[1]
            d_idx, d_val = d[0], d[1]
            if r_idx < d_idx:
                r[0] += n
                r_lst.append(r)
            else:
                d[0] += n
                d_lst.append(d)
                
        return "Radiant" if len(r_lst) > 0 else "Dire"
        

File: 649*dota2-senate/solution.java
class Solution {
    
    class Tuple {
        int idx;
        char c;
        public Tuple(int idx, char c) {
            this.idx = idx;
            this.c = c;
        }
    }
    
    public String predictPartyVictory(String senate) {
        LinkedList<Tuple> r_lst = new LinkedList<>();
        LinkedList<Tuple> d_lst = new LinkedList<>();
        int n = senate.length();
        
        for (int i = 0; i < senate.length(); i++) {
            if (senate.charAt(i) == 'R') {
                r_lst.addLast(new Tuple(i, 'R'));
            } else {
                d_lst.addLast(new Tuple(i, 'D'));
            }
        }

        while (r_lst.size() > 0 && d_lst.size() > 0) {
            Tuple r = r_lst.get(0);
            Tuple d = d_lst.get(0);
            if (r.idx < d.idx) {
                r.idx += n;
                r_lst.addLast(r);
            } else {
                d.idx += n;
                d_lst.addLast(d);
            }
            r_lst.removeFirst();
            d_lst.removeFirst();
        }
        
        if (r_lst.size() > 0) {
            return "Radiant";
        }
        return "Dire";
    }
}
